---
title: "Manuscript"
author: "James Choi"
date: 'Last compiled: `r Sys.Date()`'
output:
  pdf_document: default
  html_document:
    keep_md: yes
  word_document: default
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(warning=FALSE, message=FALSE, fig.align='center', tidy.opts=list(width.cutoff=80), tidy=TRUE, results = 'hold')
knitr::opts_knit$set(root.dir = 'D:/cns-trauma-seq/')
```


# cns-trauma-seq Notebook

Single cell sequencing resource of data from central nervous system traumatic injury.


## 2022-08-29

Workflow for this study and single cell model were separated into two tasks where the first task was to preprocess sequencing data (e.g. `.fastq`/`.fq` files) into gene-count numeric matrices and the second task was to perform cluster analysis and differential expression tests. 

One of the first issues is that I only have access to `.bam` and not `.fastq` files for certain samples. 10X Genomics has a software tool `bamtofastq` that converts 10X Genomics `.bam` files back into `.fastq` so that they can be used as input for analysis re-runs. I specifically need the raw `.bam` files, and not the files decompressed from `.sra`, due to [a known issue](https://support.10xgenomics.com/docs/bamtofastq#issues).

I requested the following data from our SRA submission (Milich2021):

  * BioProject: PRJNA682392
  * https://www.ncbi.nlm.nih.gov/Traces/study/?acc=PRJNA682392&o=acc_s%3Aa
  * `data/Milich2021_SRR_Acc_List.txt`

I saved the *Accession List* from under *Download* column of the table under the *Select* tab.

By continuing to explore the data deposited at SRA, I saw the data available like so: https://trace.ncbi.nlm.nih.gov/Traces/index.html?view=run_browser&acc=SRR13192645&display=data-access

The `.sra` and `.bam` file format options were found here.

So I needed a method to take these accessions and get the full list of download links. Using WSL2 on a Windows machine, I set up a `conda` environment named `cns-trauma-seq`. I installed the tool [`ffq` developed by Pachter Lab at UC Berkeley](https://github.com/pachterlab/ffq) which allows me to access SRA metadata using only accession codes. 

I confirmed that `ffq` is working properly using SRR13192645.

```{}
ffq SRR13192645 | head
[2022-08-29 15:34:27,261]    INFO Parsing run SRR13192645
{
    "SRR13192645": {
        "accession": "SRR13192645",
        "experiment": "SRX9627035",
        "study": "SRP295673",
        "sample": "SRS7828004",
        "title": "NextSeq 500 paired end sequencing; GSM4955359: uninj_sample1; Mus musculus; RNA-Seq",
        "attributes": {
            "assembly": "GRCm38",
            "ENA-FIRST-PUBLIC": "2021-03-02",
```