---
title: "cns-trauma-seq notebook"
author: "James Choi"
date: 'Last compiled: `r Sys.Date()`'
output:
  html_document:
    keep_md: yes
  word_document: default
  pdf_document: default
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(warning=FALSE, message=FALSE, fig.align='center', tidy.opts=list(width.cutoff=80), tidy=TRUE, results = 'hold', eval = FALSE)
knitr::opts_knit$set(root.dir = 'D:/cns-trauma-seq/')
```


## About

Single cell sequencing resource of data from central nervous system traumatic injury.


## 2022-08-29

Workflow for this study and single cell model were separated into two tasks where the first task was to preprocess sequencing data (e.g. `.fastq`/`.fq` files) into gene-count numeric matrices and the second task was to perform cluster analysis and differential expression tests. 

One of the first issues is that I only have access to `.bam` and not `.fastq` files for certain samples. 10X Genomics has a software tool `bamtofastq` that converts 10X Genomics `.bam` files back into `.fastq` so that they can be used as input for analysis re-runs. I specifically need the raw `.bam` files, and not the files decompressed from `.sra`, due to [a known issue](https://support.10xgenomics.com/docs/bamtofastq#issues). To use these tools I'll need to use the Pegasus cluster. 

I requested the following data from our SRA submission (Milich2021):

  * BioProject: PRJNA682392
  * https://www.ncbi.nlm.nih.gov/Traces/study/?acc=PRJNA682392&o=acc_s%3Aa
  * `data/Milich2021_SRR_Acc_List.txt`

I saved the *Accession List* from under *Download* column of the table under the *Select* tab.

By continuing to explore the data deposited at SRA, I saw the data available like so: https://trace.ncbi.nlm.nih.gov/Traces/index.html?view=run_browser&acc=SRR13192645&display=data-access

The `.sra` and `.bam` file format options were found here.

So I needed a method to take these accessions and get the full list of download links. Using WSL2 on a Windows machine, I set up a `conda` environment named `cns-trauma-seq`. I installed the tool [`ffq` developed by Pachter Lab at UC Berkeley](https://github.com/pachterlab/ffq) which allows me to access SRA metadata using only accession codes. 

I confirmed that `ffq` is working properly using SRR13192645. The info I needed was under `"{"SRR13192645":"files":{ftp:["url":...]}}`. 

```{}
>> ffq SRR13192645 | head
[2022-08-29 15:34:27,261]    INFO Parsing run SRR13192645
{
    "SRR13192645": {
        "accession": "SRR13192645",
        "experiment": "SRX9627035",
        "study": "SRP295673",
        "sample": "SRS7828004",
        "title": "NextSeq 500 paired end sequencing; GSM4955359: uninj_sample1; Mus musculus; RNA-Seq",
        "attributes": {
            "assembly": "GRCm38",
            "ENA-FIRST-PUBLIC": "2021-03-02",
            
>> ffq -o ffq-output.json SRR13192645 SRR13192646 SRR13192647 SRR13192648 SRR13192649 SRR13192650 SRR13192651 SRR13192652 SRR13192653 SRR13192654
```


I will eventually need to reorganize the `json` output into a batch process command. I could do this programmatically or by hand but first I need to download a test dataset. 

`ssh` to Pegasus and tested `wget https://sra-pub-src-2.s3.amazonaws.com/SRR13192645/uninj_sample1_possorted_genome_bam.bam.1`, which was ETA of 20 minutes. For all 10 samples, that's just over 3 hours. First I parse `ffq-output.json` to generate a `txt` file of just links.

```{r eval=TRUE}
links <- jsonlite::fromJSON(txt = 'ffq-output.json')
links <- lapply(links, function(x) x[['files']][['ftp']])
links <- sapply(links, function(x) x$url[x$filetype == 'bam'][1])
tmp.names <- sapply(strsplit(links, '/'), function(x) rev(x)[])
writeLines(text = links, con = 'milich2021-ftp-urls.txt')
# writeLines(text = paste0('curl -L ', links, ' -o ', tmp.names), con = 'milich2021-ftp-urls.txt')
```


## 2022-08-31

Download of bams was successful. Each bam was between 35-40Gb. 

Next was test `bamtofastq` tool from `cellranger` by 10X Genomics. I used an interactive job session on Pegasus. I followed the instructions here: https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/6.0/installation

I'm downloading `cellranger-6.0.2` because that is the version Dr Griswold at HIHG used for the aged/proliferation SCI scRNAseq study. That way, I can re-align two Milich2021 samples for the aged/proliferation study and I can test-run `bamtofastq`. 

In pegasus interactive mode:

```
# Download cellranger binary
$ curl -o cellranger-6.0.2.tar.gz "https://cf.10xgenomics.com/releases/cell-exp/cellranger-6.0.2.tar.gz?Expires=1662008119&Policy=eyJTdGF0ZW1lbnQiOlt7IlJlc291cmNlIjoiaHR0cHM6Ly9jZi4xMHhnZW5vbWljcy5jb20vcmVsZWFzZXMvY2VsbC1leHAvY2VsbHJhbmdlci02LjAuMi50YXIuZ3oiLCJDb25kaXRpb24iOnsiRGF0ZUxlc3NUaGFuIjp7IkFXUzpFcG9jaFRpbWUiOjE2NjIwMDgxMTl9fX1dfQ__&Signature=bdlpEB~od9rl6LMWCIZqvP-ESlKuyg~Upc45K0nQlHgprSBPviOZWKj6uuRA4UZK4IjQfqR94TFmgdkztg-Tv8kV3eFK0Cojyrxmokhqrqwnp0Zr5Rj4XjU66Z8jlQEtQfhjdoBfkqu34J7ZBreknFLxs9XEzygPMxew-~CwIGdVzNaIqVw9E0Vm0rAdQD44gEFdAcEgJcjyccBHJZuHD7yddfi7R58osZ~0LDTq4eaGNlJQaCTbElLgEPKoRolKWqBsOMwU1hn28vKUZjcX~h3XJyYzovqZahavghQ7tXYf6-epzB~vGjY43XlsvMFFSCbGSxI5gertXfWhg4ToLg__&Key-Pair-Id=APKAI7S6A5RYOXBWRPDA"

$ tar -xvzf cellranger-6.0.2.tar.gz

# Download reference data files (genome annotations, etc.)
$ curl -O https://cf.10xgenomics.com/supp/cell-exp/refdata-gex-mm10-2020-A.tar.gz

$ tar -xzvf refdata-gex-GRCh38-2020-A.tar.gz

# Also added to PATH in .bashrc
$ export PATH=/opt/cellranger-6.0.2:$PATH
$ cellranger sitecheck > sitecheck.txt
$ cellranger testrun --id=tiny

$ cd cellranger-6.0.2
$ cd /scratch/projects/lemmon/jsc228/cns-trauma-seq
$ /nethome/jsc228/cellranger-6.0.2/lib/bin/bamtofastq --help

bamtofastq v1.3.1
10x Genomics BAM to FASTQ converter.

    Tool for converting 10x BAMs produced by Cell Ranger or Long Ranger back to
    FASTQ files that can be used as inputs to re-run analysis. The FASTQ files
    emitted by the tool should contain the same set of sequences that were
    input to the original pipeline run, although the order will not be
    preserved.  The FASTQs will be emitted into a directory structure that is
    compatible with the directories created by the 'mkfastq' tool.

    10x BAMs produced by Long Ranger v2.1+ and Cell Ranger v1.2+ contain header
    fields that permit automatic conversion to the correct FASTQ sequences.

    Older 10x pipelines require one of the arguments listed below to indicate
    which pipeline created the BAM.
    
$ export PATH=/nethome/jsc228/cellranger-6.0.2/lib/bin/bamtofastq:$PATH

$ /nethome/jsc228/cellranger-6.0.2/lib/bin/bamtofastq --traceback data/bams/3dpi_sample1_possorted_genome_bam.bam.1 test
bamtofastq v1.3.1
Args { arg_bam: "data/bams/3dpi_sample1_possorted_genome_bam.bam.1", arg_output_path: "test", flag_nthreads: 4, flag_locus: None, flag_bx_list: None, flag_reads_per_fastq: 50000000, flag_gemcode: false, flag_lr20: false, flag_cr11: false, flag_traceback: true }
entry
```

It looks like `bamtofastq` is at least starting without installation-related errors. Depending on the resulting `fastq` files, I will have to put together a batch script for Pegasus. On second thought, I will just test the batch run. Script located in newly maked dir `lsf/pegasus-bsub_*`. Run started at 4:10pm. Finished at least a few minutes before 6:00pm. Next wrote a short lsf script to run `cellranger count` on the newly generated `.fastq.gz` files. 

The following worked the first time I ran it:

```{}
$ cellranger count \
  --id=Lindsay-Milich-14 \
  --transcriptome=/scratch/projects/lemmon/jsc228/cns-trauma-seq/refdata-gex-mm10-2020-A \
  --fastqs=/scratch/projects/lemmon/jsc228/cns-trauma-seq/data/bamtofastq/3dpi_sample1_possorted_genome_bam.bam.1/Lindsay-Milich-14_MissingLibrary_1_HV5CFBGX7
```

There are other options available for `cellranger count` but right now I'm too lazy. I'm going to hope that the develops took care of any info leaks. Batch job script `lsf/pegasus-bsub_milich2021-cellranger-count.sh`. For this job script, I submitted to the `bigmem` queue and I derived a working set of job resource request parameters.


## 2022-09-01

`cellranger count` ran successfully for `3dpi_sample1_possorted_genome_bam.bam.1` but not for `3dpi_sample1_possorted_genome_bam.bam.2`. No `outs/` directory was made and I got the following error message: 

```
2022-09-02 13:46:27 [runtime] (failed)          ID.Lindsay-Milich-3.SC_RNA_COUNTER_CS.SC_MULTI_CORE.MULTI_GEM_WELL_PROCESSOR.COUNT_GEM_WELL_PROCESSOR._BASIC_SC_RNA_COUNTER._MATRIX_COMPUTER.MAKE_SHARD

[error] Pipestance failed. Error log at:
Lindsay-Milich-3/SC_RNA_COUNTER_CS/SC_MULTI_CORE/MULTI_GEM_WELL_PROCESSOR/COUNT_GEM_WELL_PROCESSOR/_BASIC_SC_RNA_COUNTER/_MATRIX_COMPUTER/MAKE_SHARD/fork0/chnk0-u5319124163/_errors

Log message:
IO error in FASTQ file '"/scratch/projects/lemmon/jsc228/cns-trauma-seq/data/bamtofastq/3dpi_sample2_possorted_genome_bam.bam.1/2019-05-22-Lindsay-Milich-3_0_1_HHTVCBGXB/bamtofastq_S1_L001_I1_001.fastq.gz"', line: 5412024: unexpected end of file

Waiting 6 seconds for UI to do final refresh.
```

There might be an issue with the `bam` file. I checked the file with the following and got a similar message:

```
$ gunzip bamtofastq_S1_L001_I1_001.fastq.gz
```

So I will run `bamtofastq` a second time to see if the output changes. 

While that is running, I will test whether the newly aligned 3dpi-sample1 will show any batch effects with the previous 3dpi-sample1.


```{r eval=FALSE}
BiocManager::install('sparseMatrixStats')
BiocManager::install('DropletUtils')
library(Seurat)
library(ggplot2)
library(dplyr)

previous <- Read10X_h5(filename = '../MiamiProject/sci_scRNAseq/data/raw_feature_bc_matrix/raw_feature_bc_matrix_3dpi_sample2.h5')
current <- Read10X_h5(filename = 'data/raw-feature-barcode-matrices/3dpi_sample1-2022_raw_feature_bc_matrix.h5')

dim(previous)
dim(current)
summary(sparseMatrixStats::rowSums2(previous))
summary(sparseMatrixStats::rowSums2(current))
summary(sparseMatrixStats::colSums2(previous))
summary(sparseMatrixStats::colSums2(current))
```

The two matrices have different dimensions - current matrix has more genes (rows) and fewer droplets (columns). Additionally, more UMIs per cell in the current sample than previous sample. This might be due to [reference genome changes as mentioned in 10X Genomics release notes:](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/4.0/release-notes)

> Recommended reference packages for human and mouse have been updated from version 3.0.0 to 2020-A: 

> Mapping rates and gene/UMI sensitivity are increased due to more comprehensive annotations and improved manual curation of genes:

Compare `cellranger` output of sequencing metrics:

Aligned by me (2022):

Number of Reads	526,850,375
Number of Short Reads Skipped	0
Valid Barcodes	98.0%
Valid UMIs	99.9%
Sequencing Saturation	74.1%
Q30 Bases in Barcode	96.4%
Q30 Bases in RNA Read	81.2%
Q30 Bases in UMI	95.8%
Reads Mapped to Genome	89.0%
Reads Mapped Confidently to Genome	86.2%
Reads Mapped Confidently to Intergenic Regions	1.9%
Reads Mapped Confidently to Intronic Regions	9.2%
Reads Mapped Confidently to Exonic Regions	75.2%
Reads Mapped Confidently to Transcriptome	73.2%
Reads Mapped Antisense to Gene	1.0%


Aligned previously by core (2019):

Number of Reads	526,850,375
Valid Barcodes	98.0%
Sequencing Saturation	73.9%
Q30 Bases in Barcode	96.4%
Q30 Bases in RNA Read	81.2%
Q30 Bases in Sample Index	96.6%
Q30 Bases in UMI	95.8%
Reads Mapped to Genome	88.5%
Reads Mapped Confidently to Genome	84.4%
Reads Mapped Confidently to Intergenic Regions	2.7%
Reads Mapped Confidently to Intronic Regions	9.5%
Reads Mapped Confidently to Exonic Regions	72.3%
Reads Mapped Confidently to Transcriptome	69.9%
Reads Mapped Antisense to Gene	1.0%


The sequencing...

```{r eval=FALSE}
ranks_previous <- DropletUtils::barcodeRanks(
  m = previous,
  lower = 200,
  fit.bounds = c(500, 3e4)
)
ranks_current <- DropletUtils::barcodeRanks(
  m = current,
  lower = 200,
  fit.bounds = c(500, 3e4)
)
drops_previous <- DropletUtils::emptyDrops(
  m = previous,
  retain = ranks_previous@metadata$knee,
  lower = ranks_previous@metadata$inflection,
  BPPARAM = BiocParallel::SnowParam(workers = 1, type = 'SOCK')
)
drops_current <- DropletUtils::emptyDrops(
  m = current,
  retain = ranks_current@metadata$knee,
  lower = ranks_current@metadata$inflection,
  BPPARAM = BiocParallel::SnowParam(workers = 1, type = 'SOCK')
)
```


## 2022-09-06

Picking up from last time. I wrote a short script to compare the `emptyDrops` outputs which is important for determining cells sequenced but also for understanding how the additional counts per droplet (since we observed that the average UMI/cell has increased)

```{r, file='scripts/quality-control-data-preprocessing.R', eval=FALSE}
library(Seurat)
library(ggplot2)
library(dplyr)

dir.create(path = 'results/quality-control-data-processing/')
dir.create(path = 'results/quality-control-data-processing/emptyDrops/')

previous <- Read10X_h5(filename = '../MiamiProject/sci_scRNAseq/data/raw_feature_bc_matrix/raw_feature_bc_matrix_3dpi_sample2.h5')
current <- Read10X_h5(filename = 'data/raw-feature-barcode-matrices/3dpi_sample1-2022_raw_feature_bc_matrix.h5')

# Since v6.0 of cellranger by 10X Genomics, GEM-barcodes with zero total UMIs are automatically removed from the count matrix before being written out to h5.

ranks_previous <- DropletUtils::barcodeRanks(
  m = previous,
  lower = 200,
  fit.bounds = c(500, 3e4)
)
drops_previous <- DropletUtils::emptyDrops(
  m = previous,
  retain = ranks_previous@metadata$knee,
  lower = ranks_previous@metadata$inflection,
  BPPARAM = BiocParallel::SnowParam(workers = 2, type = 'SOCK')
)
previous_results <- cbind(drops_previous, ranks_previous)
rm(ranks_previous, drops_previous)

ranks_current <- DropletUtils::barcodeRanks(
  m = current,
  lower = 200,
  fit.bounds = c(500, 3e4)
)
drops_current <- DropletUtils::emptyDrops(
  m = current,
  retain = ranks_current@metadata$knee,
  lower = ranks_current@metadata$inflection,
  BPPARAM = BiocParallel::SnowParam(workers = 2, type = 'SOCK')
)
current_results <- cbind(drops_current, ranks_current)
rm(ranks_current, drops_current)

all_results <- list(
  previous_results = previous_results,
  current_results = current_results
)
saveRDS(all_results, file = 'results/quality-control-data-processing/emptyDrops/emptyDrops-outs_barcodeRanks-outs.rds')
```


## 2022-09-07

Edited `quality-control-data-preprocessing.R` to run correctly.

## 2022-09-09

Goal today is to do side-by-side comparison of previously aligned and recently aligned 3dpi-sample1 datasets.

Noticed the distributions were wildly different. Looking back through previous work, turns out `quality-control-data-preprocessing.R` had imported `3dpi_sample2` instead of `3dpi_sample1`. This was before I wrote the script below:


```{r}
library(Seurat)
library(ggplot2)
library(dplyr)

dir.create(path = 'results/quality-control-data-processing/')

previous <- Read10X_h5(filename = '../MiamiProject/sci_scRNAseq/data/raw_feature_bc_matrix/raw_feature_bc_matrix_3dpi_sample1.h5')
current <- Read10X_h5(filename = 'data/raw-feature-barcode-matrices/3dpi_sample1-2022_raw_feature_bc_matrix.h5')
all_results <- readRDS(file = 'results/quality-control-data-processing/emptyDrops/emptyDrops-outs_barcodeRanks-outs.rds')

previous <- previous[, which(all_results$previous_results$FDR < 0.05)]
current <- current[, which(all_results$current_results$FDR < 0.05)]
dim(current)
dim(previous)

previous <- CreateSeuratObject(counts = previous, project = 'previous')
current <- CreateSeuratObject(counts = current, project = 'current')

previous <- PercentageFeatureSet(previous, pattern = '^mt-', col.name = 'percent_mt')
current <- PercentageFeatureSet(current, pattern = '^mt-', col.name = 'percent_mt')

p <- rbind(previous@meta.data, current@meta.data) %>% 
  reshape2::melt(id.vars = c('orig.ident')) %>% 
  ggplot(mapping = aes(x = orig.ident, y = value)) + 
  geom_violin(scale = 'width') + 
  facet_wrap(. ~ variable, scales = 'free_y') +
  theme_bw() +
  labs(title = 'cellranger count v2.0 vs v6.1', subtitle = 'previous = Milich2019, current = Choi2022')
ggsave(filename = 'results/quality-control-data-processing/cellranger-rerun-comparison_3dpi-sample1.tiff', plot = p, device = 'tiff', height = 3, width = 4.5, dpi = 320)
```

The violin shows us that the current version of `cellranger` has a nearly identical distribution of `nCount_RNA` and `nFeature_RNA`. However, the previous version has wider tails towards the lower values, indicating that previous data has more cells with fewer UMIs and genes. 

```{r}
merged <- merge(previous, current)
merged <- merged %>% 
  NormalizeData() %>% 
  FindVariableFeatures(nfeatures = 2000) %>% 
  ScaleData() %>% 
  RunPCA() %>% 
  FindNeighbors(dims = 1:10) %>% 
  FindClusters() %>% 
  RunUMAP(dims = 1:10)
```


I added to `quality-control-data-processing.R` to include the gene-level QC computations as well as provide visuals for the QC results i.e. barcode-rank graphs for the previous cell-calling done.


## 2022-10-17

I submitted a Pegasus job on 2022-10-17 to run `bamtofastq` from `cellranger` on all of the 2021 SCI study samples.

Said job:

```
#!/bin/bash
#BSUB -J cns-trauma-seq_bamtofastq
#BSUB -P lemmon
#BSUB -o %J.out
#BSUB -e %J.err
#BSUB -W 6:00
#BSUB -q general
#BSUB -n 8
#BSUB -B
#BSUB -N
#BSUB -u jsc228@miami.edu

####################################################
# Set directory variables and conda env
####################################################

export PATH=/nethome/jsc228/cellranger-6.0.2:$PATH
cd /scratch/projects/lemmon/jsc228/cns-trauma-seq
# BAM_DIR=data/bams/
# FASTQ_DIR=data/bamtofastq/
# SAMPLE_NAMES=$(cat milich2021-ftp-urls.txt | rev | cut -d "/" -f 1 | rev)
# if [! -d "$BAM_PATH" ]; then mkdir "$BAM_PATH"; fi


####################################################
# Job
####################################################

/nethome/jsc228/cellranger-6.0.2/lib/bin/bamtofastq \
  --traceback \
  --nthreads=8 \
  data/bams/uninj_sample1_possorted_genome_bam.bam.1 \
  data/bamtofastq/uninj_sample1_possorted_genome_bam.bam.1/

/nethome/jsc228/cellranger-6.0.2/lib/bin/bamtofastq \
  --traceback \
  --nthreads=8 \
  data/bams/uninj_sample2_possorted_genome_bam.bam.1 \
  data/bamtofastq/uninj_sample2_possorted_genome_bam.bam.1/

/nethome/jsc228/cellranger-6.0.2/lib/bin/bamtofastq \
  --traceback \
  --nthreads=8 \
  data/bams/uninj_sample3_possorted_genome_bam.bam.1 \
  data/bamtofastq/uninj_sample3_possorted_genome_bam.bam.1/

/nethome/jsc228/cellranger-6.0.2/lib/bin/bamtofastq \
  --traceback \
  --nthreads=8 \
  data/bams/1dpi_sample1_possorted_genome_bam.bam.1 \
  data/bamtofastq/1dpi_sample1_possorted_genome_bam.bam.1/

/nethome/jsc228/cellranger-6.0.2/lib/bin/bamtofastq \
  --traceback \
  --nthreads=8 \
  data/bams/1dpi_sample2_possorted_genome_bam.bam.1 \
  data/bamtofastq/1dpi_sample2_possorted_genome_bam.bam.1/

/nethome/jsc228/cellranger-6.0.2/lib/bin/bamtofastq \
  --traceback \
  --nthreads=8 \
  data/bams/1dpi_sample3_possorted_genome_bam.bam.1 \
  data/bamtofastq/1dpi_sample3_possorted_genome_bam.bam.1/

/nethome/jsc228/cellranger-6.0.2/lib/bin/bamtofastq \
  --traceback \
  --nthreads=8 \
  data/bams/3dpi_sample1_possorted_genome_bam.bam.1 \
  data/bamtofastq/3dpi_sample1_possorted_genome_bam.bam.1/

/nethome/jsc228/cellranger-6.0.2/lib/bin/bamtofastq \
  --traceback \
  --nthreads=8 \
  data/bams/3dpi_sample2_possorted_genome_bam.bam.1 \
  data/bamtofastq/3dpi_sample2_possorted_genome_bam.bam.1/

/nethome/jsc228/cellranger-6.0.2/lib/bin/bamtofastq \
  --traceback \
  --nthreads=8 \
  data/bams/7dpi_sample1_possorted_genome_bam.bam.1 \
  data/bamtofastq/7dpi_sample1_possorted_genome_bam.bam.1/

/nethome/jsc228/cellranger-6.0.2/lib/bin/bamtofastq \
  --traceback \
  --nthreads=8 \
  data/bams/7dpi_sample2_possorted_genome_bam.bam.1 \
  data/bamtofastq/7dpi_sample2_possorted_genome_bam.bam.1/

####################################################
# Script End
####################################################
```

I took the easy way out and didn't bother trying to run a for-loop on the process because I'm not that well versed in bash scripting. It's only 10 samples. However, when I need to run all other samples for the larger atlas compilation, I will have to re-think this. Even if all the public samples I collect are sequenced via 10X, demultiplexing may be different depending on file type availability. `fastq`'s will hopefully be easier to align later than the `bams` I've been working with. I should upload the original `fastq` files onto GEO.

I started the run at approximately 7:00pm on Friday 2022-10-14, and I received an LSF message at 12:30am Sunday 2022-10-16. The run generated `fastq`'s for (I think) 8 of 10 samples (excluding the 3dpi samples I ran previously).


I started running `cellranger count` for all of the newly generated `fastq` files. There were no error messages related to counting, rather error messages about already existing files/directories. As such, I deleted all pegasus-generated out/err messages (i.e. `.out` and `.err` files).


Updated script for `cellranger count`:

```
#!/bin/bash
#BSUB -J cns-trauma-seq_cellranger-count
#BSUB -P lemmon
#BSUB -o %J.out
#BSUB -e %J.err
#BSUB -W 10:00
#BSUB -q bigmem
#BSUB -n 10
#BSUB -R "rusage[mem=4096]"
#BSUB -B
#BSUB -N
#BSUB -u jsc228@miami.edu

####################################################
# Set directory variables and conda env
####################################################

export PATH=/nethome/jsc228/cellranger-6.0.2:$PATH

# ! NOTE: in-run $HOME is not cns-trauma-seq/
cd /scratch/projects/lemmon/jsc228/cns-trauma-seq/data
mkdir cellranger
cd cellranger

####################################################
# Job
####################################################

# 1dpi sample1
cellranger count \
  --id=Lindsay-Milich \
  --transcriptome=/scratch/projects/lemmon/jsc228/cns-trauma-seq/refdata-gex-mm10-2020-A \
  --fastqs=/scratch/projects/lemmon/jsc228/cns-trauma-seq/data/bamtofastq/1dpi_sample1_possorted_genome_bam.bam.1/Lindsay-Milich_MissingLibrary_1_HMVM5BGX7
# 1dpi sample 2
cellranger count \
  --id=2019-05-21-Lindsay-Milich-1 \
  --transcriptome=/scratch/projects/lemmon/jsc228/cns-trauma-seq/refdata-gex-mm10-2020-A \
  --fastqs=/scratch/projects/lemmon/jsc228/cns-trauma-seq/data/bamtofastq/1dpi_sample2_possorted_genome_bam.bam.1/2019-05-21-Lindsay-Milich-1_0_1_HGK7YBGXB
# 1dpi sample 3
cellranger count \
  --id=Lee-1-2-2020-09-22 \
  --transcriptome=/scratch/projects/lemmon/jsc228/cns-trauma-seq/refdata-gex-mm10-2020-A \
  --fastqs=/scratch/projects/lemmon/jsc228/cns-trauma-seq/data/bamtofastq/1dpi_sample3_possorted_genome_bam.bam.1/Lee-1-2-2020-09-22_0_1_HNTGLBGXF
# 3dpi sample 1
cellranger count \
  --id=Lindsay-Milich-14 \
  --transcriptome=/scratch/projects/lemmon/jsc228/cns-trauma-seq/refdata-gex-mm10-2020-A \
  --fastqs=/scratch/projects/lemmon/jsc228/cns-trauma-seq/data/bamtofastq/3dpi_sample1_possorted_genome_bam.bam.1/Lindsay-Milich-14_MissingLibrary_1_HV5CFBGX7
# 3dpi sample 2
cellranger count \
  --id=2019-05-22-Lindsay-Milich-3 \
  --transcriptome=/scratch/projects/lemmon/jsc228/cns-trauma-seq/refdata-gex-mm10-2020-A \
  --fastqs=/scratch/projects/lemmon/jsc228/cns-trauma-seq/data/bamtofastq/3dpi_sample2_possorted_genome_bam.bam.1/2019-05-22-Lindsay-Milich-3_0_1_HHTVCBGXB
# 7dpi sample 1
cellranger count \
  --id=Lindsay-Milich-78 \
  --transcriptome=/scratch/projects/lemmon/jsc228/cns-trauma-seq/refdata-gex-mm10-2020-A \
  --fastqs=/scratch/projects/lemmon/jsc228/cns-trauma-seq/data/bamtofastq/7dpi_sample1_possorted_genome_bam.bam.1/Lindsay-Milich-78_MissingLibrary_1_HV5F5BGX7
# 7dpi sample 2
cellranger count \
  --id=2019-05-23-Lindsay-Milich-7 \
  --transcriptome=/scratch/projects/lemmon/jsc228/cns-trauma-seq/refdata-gex-mm10-2020-A \
  --fastqs=/scratch/projects/lemmon/jsc228/cns-trauma-seq/data/bamtofastq/7dpi_sample2_possorted_genome_bam.bam.1/2019-05-23-Lindsay-Milich-7_0_1_HHTVNBGXB
# uninj sample1
cellranger count \
  --id=Lindsay-Milich-95 \
  --transcriptome=/scratch/projects/lemmon/jsc228/cns-trauma-seq/refdata-gex-mm10-2020-A \
  --fastqs=/scratch/projects/lemmon/jsc228/cns-trauma-seq/data/bamtofastq/uninj_sample1_possorted_genome_bam.bam.1/Lindsay-Milich-95_MissingLibrary_1_HMVKHBGX7
# uninj sample2
cellranger count \
  --id=Lindsay-Milich-2-1 \
  --transcriptome=/scratch/projects/lemmon/jsc228/cns-trauma-seq/refdata-gex-mm10-2020-A \
  --fastqs=/scratch/projects/lemmon/jsc228/cns-trauma-seq/data/bamtofastq/uninj_sample2_possorted_genome_bam.bam.1/Lindsay-Milich-2-1_0_1_HM22WBGX9
# uninj sample3
cellranger count \
  --id=Lee-U-2-2020-09-23 \
  --transcriptome=/scratch/projects/lemmon/jsc228/cns-trauma-seq/refdata-gex-mm10-2020-A \
  --fastqs=/scratch/projects/lemmon/jsc228/cns-trauma-seq/data/bamtofastq/uninj_sample3_possorted_genome_bam.bam.1/Lee-U-2-2020-09-23_0_1_HTYVNBGXG


####################################################
# Script End
####################################################
```

I had to re-adjust my `.gitignore` file in such a way so that error messages related to non-existing directories would be prevented. I did this by manually specifying sub-directories in the `results` folder so that a reader of the git repository would know approximately which analyses were performed.

I double checked to see whether specific input/output files are needed for downstream RNA velocity analysis, specificalyl scVelo. All that is needed is the resulting `.bam` file, so I am running `cellranger count` with nearly default settings.


